File: crates/voidx-tools/src/file_ops.rs
//! File operation tools — read, write, edit.
//!
//! Ported from `src/voidx/tools/file_ops.py`.

use crate::base::{resolve_safe, Tool, ToolContext, ToolResult};
use crate::error::ToolError;
use crate::schema::model_to_json_schema;
use async_trait::async_trait;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use std::path::PathBuf;

// ── Shared helpers ─────────────────────────────────────────────────────────

fn read_file_content(path: &std::path::Path) -> Result<String, ToolError> {
    if !path.exists() {
        return Err(ToolError::Other(format!("File not found: {}", path.display())));
    }
    if !path.is_file() {
        return Err(ToolError::Other(format!("Not a file: {}", path.display())));
    }
    Ok(std::fs::read_to_string(path)?)
}

fn check_in_workspace(
    workspace: &std::path::Path,
    file_path: &str,
    extra_paths: &[PathBuf],
) -> Result<PathBuf, ToolError> {
    resolve_safe(workspace, file_path, extra_paths)
        .ok_or_else(|| ToolError::SandboxViolation(file_path.to_string()))
}

// ── FileReadTool ───────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct FileReadInput {
    /// The absolute path to the file to read
    pub file_path: String,
    /// The line number to start reading from (0-indexed, optional)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub offset: Option<usize>,
    /// The number of lines to read (optional, max 2000)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub limit: Option<usize>,
}

pub struct FileReadTool;

#[async_trait]
impl Tool for FileReadTool {
    fn id(&self) -> &'static str {
        "file_read"
    }

    fn description(&self) -> &'static str {
        "Read a file from the workspace. Returns content with line numbers."
    }

    fn parameters_schema(&self) -> serde_json::Value {
        model_to_json_schema::<FileReadInput>()
    }

    async fn execute(
        &self,
        args: serde_json::Value,
        ctx: &ToolContext,
    ) -> Result<ToolResult, ToolError> {
        let inp: FileReadInput = serde_json::from_value(args)
            .map_err(|e| ToolError::InvalidArgs(e.to_string()))?;

        let path = check_in_workspace(&ctx.workspace, &inp.file_path, &ctx.sandbox_extra_paths)?;
        let content = read_file_content(&path)?;

        let offset = inp.offset.unwrap_or(0);
        let limit = inp.limit.unwrap_or(2000).min(2000);

        let lines: Vec<&str> = content.lines().skip(offset).take(limit).collect();
        let output = lines
            .iter()
            .enumerate()
            .map(|(i, line)| format!("{:>6}\t{}", offset + i + 1, line))
            .collect::<Vec<_>>()
            .join("\n");

        Ok(ToolResult::new(output)
            .with_title(format!("File: {}", inp.file_path))
            .with_metadata(serde_json::json!({
                "file_path": inp.file_path,
                "total_lines": content.lines().count(),
                "offset": offset,
                "limit": limit,
            })))
    }
}

// ── FileWriteTool ──────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct FileWriteInput {
    /// The absolute path to the file to write
    pub file_path: String,
    /// The content to write to the file
    pub content: String,
}

pub struct FileWriteTool;

#[async_trait]
impl Tool for FileWriteTool {
    fn id(&self) -> &'static str {
        "file_write"
    }

    fn description(&self) -> &'static str {
        "Write a file to the workspace. Overwrites existing files."
    }

    fn parameters_schema(&self) -> serde_json::Value {
        model_to_json_schema::<FileWriteInput>()
    }

    async fn execute(
        &self,
        args: serde_json::Value,
        ctx: &ToolContext,
    ) -> Result<ToolResult, ToolError> {
        let inp: FileWriteInput = serde_json::from_value(args)
            .map_err(|e| ToolError::InvalidArgs(e.to_string()))?;

        let path = check_in_workspace(&ctx.workspace, &inp.file_path, &ctx.sandbox_extra_paths)?;

        // Ensure parent directory exists
        if let Some(parent) = path.parent() {
            std::fs::create_dir_all(parent)?;
        }

        let exists = path.exists();
        std::fs::write(&path, &inp.content)?;

        let line_count = inp.content.lines().count();
        let action = if exists { "Updated" } else { "Created" };

        Ok(ToolResult::new(format!(
            "{action} {} ({} lines)",
            inp.file_path, line_count
        ))
        .with_title(format!("Write: {}", inp.file_path))
        .with_metadata(serde_json::json!({
            "file_path": inp.file_path,
            "lines": line_count,
            "existed": exists,
        })))
    }
}

// ── FileEditTool ───────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct FileEditInput {
    /// The absolute path to the file to modify
    pub file_path: String,
    /// The text to replace
    pub old_string: String,
    /// The text to replace it with (must differ from old_string)
    pub new_string: String,
    /// Replace all occurrences (default: false)
    #[serde(default)]
    pub replace_all: bool,
}

pub struct FileEditTool;

#[async_trait]
impl Tool for FileEditTool {
    fn id(&self) -> &'static str {
        "file_edit"
    }

    fn description(&self) -> &'static str {
        "Perform exact string replacement in a file."
    }

    fn parameters_schema(&self) -> serde_json::Value {
        model_to_json_schema::<FileEditInput>()
    }

    async fn execute(
        &self,
        args: serde_json::Value,
        ctx: &ToolContext,
    ) -> Result<ToolResult, ToolError> {
        let inp: FileEditInput = serde_json::from_value(args)
            .map_err(|e| ToolError::InvalidArgs(e.to_string()))?;

        if inp.old_string == inp.new_string {
            return Err(ToolError::InvalidArgs(
                "old_string and new_string must differ".to_string(),
            ));
        }

        let path = check_in_workspace(&ctx.workspace, &inp.file_path, &ctx.sandbox_extra_paths)?;
        let original = read_file_content(&path)?;

        if inp.old_string.is_empty() {
            return Err(ToolError::InvalidArgs("old_string must not be empty".to_string()));
        }

        let occurrences = original.matches(&inp.old_string).count();
        if occurrences == 0 {
            return Err(ToolError::Other(format!(
                "old_string not found in {}",
                inp.file_path
            )));
        }

        if occurrences > 1 && !inp.replace_all {
            return Err(ToolError::Other(format!(
                "old_string found {occurrences} times; set replace_all=true to replace all, or make old_string more specific"
            )));
        }

        let modified = if inp.replace_all {
            original.replace(&inp.old_string, &inp.new_string)
        } else {
            original.replacen(&inp.old_string, &inp.new_string, 1)
        };

        std::fs::write(&path, &modified)?;

        // Generate unified diff
        let diff = similar::TextDiff::from_lines(&original, &modified);
        let diff_str = diff
            .unified_diff()
            .context_radius(3)
            .header(&inp.file_path, &inp.file_path)
            .to_string();

        let replaced = if inp.replace_all { occurrences } else { 1 };

        Ok(ToolResult::new(format!(
            "Applied edit to {}: {} replacement(s)",
            inp.file_path, replaced
        ))
        .with_title(format!("Edit: {}", inp.file_path))
        .with_diff(diff_str)
        .with_metadata(serde_json::json!({
            "file_path": inp.file_path,
            "replacements": replaced,
        })))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Write;

    fn temp_workspace() -> tempfile::TempDir {
        tempfile::tempdir().unwrap()
    }

    fn create_file(dir: &std::path::Path, name: &str, content: &str) {
        let path = dir.join(name);
        if let Some(parent) = path.parent() {
            std::fs::create_dir_all(parent).unwrap();
        }
        let mut f = std::fs::File::create(path).unwrap();
        f.write_all(content.as_bytes()).unwrap();
    }

    #[tokio::test]
    async fn test_file_read() {
        let ws = temp_workspace();
        create_file(ws.path(), "test.txt", "line1\nline2\nline3\n");

        let tool = FileReadTool;
        let ctx = ToolContext {
            workspace: ws.path().to_path_buf(),
            ..Default::default()
        };

        let result = tool
            .execute(
                serde_json::json!({"file_path": "test.txt"}),
                &ctx,
            )
            .await
            .unwrap();

        assert!(result.output.contains("line1"));
        assert!(result.output.contains("line2"));
    }

    #[tokio::test]
    async fn test_file_write() {
        let ws = temp_workspace();

        let tool = FileWriteTool;
        let ctx = ToolContext {
            workspace: ws.path().to_path_buf(),
            ..Default::default()
        };

        let result = tool
            .execute(
                serde_json::json!({"file_path": "new.txt", "content": "hello world"}),
                &ctx,
            )
            .await
            .unwrap();

        assert!(result.output.contains("Created"));
        let content = std::fs::read_to_string(ws.path().join("new.txt")).unwrap();
        assert_eq!(content, "hello world");
    }

    #[tokio::test]
    async fn test_file_edit() {
        let ws = temp_workspace();
        create_file(ws.path(), "edit.txt", "Hello Alice\nHello Bob\n");

        let tool = FileEditTool;
        let ctx = ToolContext {
            workspace: ws.path().to_path_buf(),
            ..Default::default()
        };

        let result = tool
            .execute(
                serde_json::json!({
                    "file_path": "edit.txt",
                    "old_string": "Alice",
                    "new_string": "Charlie",
                    "replace_all": false,
                }),
                &ctx,
            )
            .await
            .unwrap();

        assert!(result.output.contains("1 replacement"));
        let content = std::fs::read_to_string(ws.path().join("edit.txt")).unwrap();
        assert!(content.contains("Charlie"));
        assert!(result.diff.is_some());
    }

    #[tokio::test]
    async fn test_file_read_outside_workspace_blocked() {
        let ws = temp_workspace();

        let tool = FileReadTool;
        let ctx = ToolContext {
            workspace: ws.path().to_path_buf(),
            ..Default::default()
        };

        let result = tool
            .execute(
                serde_json::json!({"file_path": "../outside.txt"}),
                &ctx,
            )
            .await;

        assert!(result.is_err());
    }
}

File: python/voidx/tools/file_ops.py
"""File operation tools — read, write, edit. Deterministic, typed I/O."""

from __future__ import annotations

from pathlib import Path

from pydantic import BaseModel, Field

from voidx.tools.base import BaseTool, model_to_json_schema, ToolContext, ToolResult, resolve_safe


def _check_staleness(ctx: ToolContext, resolved: Path) -> str | None:
    """Return error message if file was modified since last read, else None."""
    key = str(resolved.resolve())
    if key not in ctx.file_mtimes:
        return None
    if not resolved.exists():
        return f"File deleted since last read: {resolved}"
    current_mtime = resolved.stat().st_mtime
    if current_mtime != ctx.file_mtimes[key]:
        return (
            f"File was modified since last read: {resolved}. "
            "Please re-read the file before editing."
        )
    return None


def _record_mtime(ctx: ToolContext, resolved: Path) -> None:
    """Record file mtime after successful read or write."""
    if resolved.exists():
        ctx.file_mtimes[str(resolved.resolve())] = resolved.stat().st_mtime


class FileReadInput(BaseModel):
    file_path: str = Field(description="Absolute or relative path to the file")
    offset: int | None = Field(default=None, description="Line number to start reading from (1-based)")
    limit: int | None = Field(default=None, description="Maximum number of lines to read")


class FileReadTool(BaseTool):
    id = "read"
    description = "Read a file. Returns content with line numbers. Use offset/limit for large files."

    def parameters_schema(self) -> dict:
        return model_to_json_schema(FileReadInput)

    async def execute(self, args: dict, ctx: ToolContext) -> ToolResult:
        inp = FileReadInput.model_validate(args)
        path = resolve_safe(ctx.workspace, inp.file_path, ctx.sandbox_extra_paths)
        if path is None:
            return ToolResult(output=f"Path traversal blocked: {inp.file_path}", metadata={"error": True})
        if not path.exists():
            return ToolResult(output=f"File not found: {inp.file_path}", metadata={"error": True})
        if path.is_dir():
            return ToolResult(output=f"Path is a directory: {inp.file_path}", metadata={"error": True})

        text = path.read_text(encoding="utf-8", errors="replace")
        if text.endswith("\n"):
            text = text[:-1]
        lines = text.split("\n")
        start = (inp.offset or 1) - 1
        if start >= len(lines):
            return ToolResult(
                title=f"Read 0 lines from {inp.file_path}",
                output=f"Offset {inp.offset} is beyond end of file (file has {len(lines)} lines).",
                metadata={"file": inp.file_path, "lines": 0, "total_lines": len(lines)},
            )
        end = start + (inp.limit or len(lines))
        sliced = lines[start:end]

        numbered = []
        for i, line in enumerate(sliced, start=start + 1):
            numbered.append(f"{i}\t{line}")

        _record_mtime(ctx, path)

        return ToolResult(
            title=f"Read {len(sliced)} lines from {inp.file_path}",
            output="\n".join(numbered),
            metadata={"file": inp.file_path, "lines": len(sliced), "total_lines": len(lines)},
        )


class FileWriteInput(BaseModel):
    file_path: str = Field(description="Path to write the file to")
    content: str = Field(description="Content to write")


class FileWriteTool(BaseTool):
    id = "write"
    description = "Write content to a file. Creates parent directories. Overwrites existing files."

    def parameters_schema(self) -> dict:
        return model_to_json_schema(FileWriteInput)

    async def execute(self, args: dict, ctx: ToolContext) -> ToolResult:
        inp = FileWriteInput.model_validate(args)
        path = resolve_safe(ctx.workspace, inp.file_path, ctx.sandbox_extra_paths)
        if path is None:
            return ToolResult(output=f"Path traversal blocked: {inp.file_path}", metadata={"error": True})
        if path.exists():
            stale = _check_staleness(ctx, path)
            if stale:
                return ToolResult(output=stale, metadata={"error": True})

        old_content = ""
        if path.exists():
            old_content = path.read_text(encoding="utf-8", errors="replace")

        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(inp.content, encoding="utf-8")
        size = len(inp.content)
        _record_mtime(ctx, path)

        from voidx.ui.output.diff import make_file_diff
        diff = make_file_diff(
            inp.file_path,
            old_content,
            inp.content,
            old_label=f"a/{inp.file_path}" if old_content else "/dev/null",
            new_label=f"b/{inp.file_path}",
        )

        return ToolResult(
            title=f"Wrote {size} bytes to {inp.file_path}",
            output=f"File written: {inp.file_path} ({size} bytes)",
            metadata={"file": inp.file_path, "size": size},
            diff=diff,
        )


class EditEntry(BaseModel):
    old_string: str = Field(description="Exact string to replace. Must exist at least once in the file.")
    new_string: str = Field(description="String to replace with")


class FileEditInput(BaseModel):
    file_path: str = Field(description="Path to edit")
    edits: list[EditEntry] = Field(description="List of edits to apply atomically. Use a single entry for one change.")


class FileEditTool(BaseTool):
    id = "edit"
    description = (
        "Replace strings in a single file atomically. Each old_string must match "
        "exactly once — provide more context if it appears multiple times. "
        "Edits apply in order — later edits see earlier results."
    )

    def parameters_schema(self) -> dict:
        return model_to_json_schema(FileEditInput)

    async def execute(self, args: dict, ctx: ToolContext) -> ToolResult:
        inp = FileEditInput.model_validate(args)
        path = resolve_safe(ctx.workspace, inp.file_path, ctx.sandbox_extra_paths)
        if path is None:
            return ToolResult(output=f"Path traversal blocked: {inp.file_path}", metadata={"error": True})
        if not path.exists():
            return ToolResult(output=f"File not found: {inp.file_path}", metadata={"error": True})

        if not inp.edits:
            return ToolResult(
                output="No edits provided. The 'edits' array must contain at least one entry.",
                metadata={"error": True},
            )

        stale = _check_staleness(ctx, path)
        if stale:
            return ToolResult(output=stale, metadata={"error": True})

        original = path.read_text(encoding="utf-8", errors="replace")
        content = original

        for i, edit in enumerate(inp.edits):
            if not edit.old_string:
                return ToolResult(output=f"Edit {i}: old_string must not be empty")
            if edit.old_string == edit.new_string:
                return ToolResult(output=f"Edit {i}: old_string and new_string must differ")

            count = content.count(edit.old_string)
            if count == 0:
                return ToolResult(
                    output=f"Edit {i}: old_string not found in {inp.file_path}",
                    metadata={"error": True},
                )
            if count > 1:
                return ToolResult(
                    output=(
                        f"Edit {i}: old_string matches {count} times in {inp.file_path}. "
                        "Provide more context to make the match unique, or use write to replace the entire file."
                    ),
                    metadata={"error": True, "match_count": count},
                )

            content = content.replace(edit.old_string, edit.new_string)

        path.write_text(content, encoding="utf-8")
        _record_mtime(ctx, path)

        from voidx.ui.output.diff import make_file_diff
        diff = make_file_diff(inp.file_path, original, content)

        return ToolResult(
            title=f"Edited {inp.file_path} ({len(inp.edits)} edits)",
            output=f"File edited: {inp.file_path} ({len(inp.edits)} replacements)\n{diff}",
            metadata={"file": inp.file_path, "replacements": len(inp.edits)},
            diff=diff,
        )
